File: shared/skills/deadline_extract.py
#!/usr/bin/env python3
"""deadline_extract — lightweight parser for `positions.deadline` (F-4 #50).

The Scout or Analyst calls this skill BEFORE inserting a position to extract
the deadline from the job description. Typical English examples:

  "apply by 2026-06-15"
  "deadline: June 15, 2026"
  "applications close on Friday"
  "expires in 30 days"

Italian input remains supported for compatibility. The parser is conservative:
it extracts only high-confidence matches (ISO date, "Month dd[, yyyy]",
"dd/mm[/yyyy]", or "expires in N days"). Otherwise it returns None; a NULL
database value is safer than an invented date.

CLI:
    python3 /app/shared/skills/deadline_extract.py < jd.txt
    cat jd.txt | python3 /app/shared/skills/deadline_extract.py
    python3 /app/shared/skills/deadline_extract.py --jd "apply by 2026-06-15"

Output: an ISO date (YYYY-MM-DD), or an empty string when none is found.
Exit code is always 0; callers only need to check whether output is non-empty.
"""
from __future__ import annotations

import argparse
import re
import sys
from datetime import date, datetime, timedelta


MONTHS = {
    "jan": 1, "january": 1, "gen": 1, "gennaio": 1,
    "feb": 2, "february": 2, "febbraio": 2,
    "mar": 3, "march": 3, "marzo": 3,
    "apr": 4, "april": 4, "aprile": 4,
    "may": 5, "maggio": 5, "mag": 5,
    "jun": 6, "june": 6, "giu": 6, "giugno": 6,
    "jul": 7, "july": 7, "lug": 7, "luglio": 7,
    "aug": 8, "august": 8, "ago": 8, "agosto": 8,
    "sep": 9, "sept": 9, "september": 9, "set": 9, "settembre": 9,
    "oct": 10, "october": 10, "ott": 10, "ottobre": 10,
    "nov": 11, "november": 11, "novembre": 11,
    "dec": 12, "december": 12, "dic": 12, "dicembre": 12,
}


def _today() -> date:
    return datetime.now().date()


def _iso(y: int, m: int, d: int) -> str | None:
    try:
        return date(y, m, d).isoformat()
    except ValueError:
        return None


def _norm(text: str) -> str:
    return re.sub(r"\s+", " ", text or "").strip()


def parse_deadline(text: str) -> str | None:
    """Ritorna ISO date string o None se nessun match a confidenza alta.

    Cerco in ordine: ISO, num/num/year, Month dd[, yyyy], "expires in N days".
    Stop al primo match ragionevole (no fuzzy ranking — meglio None che data
    sbagliata).
    """
    t = _norm(text).lower()
    if not t:
        return None

    today = _today()

    # 1. ISO YYYY-MM-DD (anche dentro "by 2026-06-15", "deadline 2026-06-15")
    m = re.search(r"\b(\d{4})-(\d{2})-(\d{2})\b", t)
    if m:
        iso = _iso(int(m.group(1)), int(m.group(2)), int(m.group(3)))
        if iso and date.fromisoformat(iso) >= today:
            return iso

    # 2. dd/mm/yyyy or dd-mm-yyyy (formato EU)
    m = re.search(r"\b(\d{1,2})[/\-](\d{1,2})[/\-](\d{2,4})\b", t)
    if m:
        d_, mo_, y_ = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if y_ < 100:
            y_ += 2000
        iso = _iso(y_, mo_, d_)
        if iso and date.fromisoformat(iso) >= today:
            return iso

    # 3. "Month dd[, yyyy]" oppure "dd Month [yyyy]" (EN + IT)
    months_alt = "|".join(MONTHS.keys())
    # "June 15, 2026" / "June 15 2026" / "June 15"
    m = re.search(rf"\b({months_alt})\s+(\d{{1,2}})(?:[,\s]+(\d{{4}}))?\b", t)
    if m:
        mo_ = MONTHS[m.group(1)]
        d_ = int(m.group(2))
        y_ = int(m.group(3)) if m.group(3) else today.year
        iso = _iso(y_, mo_, d_)
        if iso:
            parsed = date.fromisoformat(iso)
            # Senza anno: se è già passato quest'anno, intendiamo l'anno prossimo
            if not m.group(3) and parsed < today:
                iso = _iso(y_ + 1, mo_, d_)
            if iso and date.fromisoformat(iso) >= today:
                return iso

    # "15 giugno 2026" / "15 June"
    m = re.search(rf"\b(\d{{1,2}})\s+({months_alt})(?:\s+(\d{{4}}))?\b", t)
    if m:
        d_ = int(m.group(1))
        mo_ = MONTHS[m.group(2)]
        y_ = int(m.group(3)) if m.group(3) else today.year
        iso = _iso(y_, mo_, d_)
        if iso:
            parsed = date.fromisoformat(iso)
            if not m.group(3) and parsed < today:
                iso = _iso(y_ + 1, mo_, d_)
            if iso and date.fromisoformat(iso) >= today:
                return iso

    # 4. "expires in N days" / "closes in N days" / "scade fra N giorni"
    m = re.search(
        r"\b(?:expires?|closes?|deadline|scade|chiude)\s+(?:in\s+|fra\s+|tra\s+|entro\s+)?(\d+)\s+(?:days?|giorni)\b",
        t,
    )
    if m:
        n = int(m.group(1))
        return (today + timedelta(days=n)).isoformat()

    # 5. Hint words senza data → skip (meglio NULL che inventato)
    return None


def main(argv):
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--jd", help="Testo JD inline (alternativa a stdin)")
    args = p.parse_args(argv)

    text = args.jd if args.jd else sys.stdin.read()
    res = parse_deadline(text) or ""
    print(res)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: agent-harness/runtime/src/parity/skills/deadline-extract.ts
/**
 * `deadline_extract.py` as a native tool (T14, ANALISTA MAIN LOOP step 5).
 *
 * The same conservative parser: the first match of each pattern in the
 * Python's order — ISO date, d/m/y, "Month dd[, yyyy]", "dd Month [yyyy]",
 * "expires in N days" — kept only if it is today or later, and an empty line
 * when nothing is certain. The regexes read as Python's `str` patterns read:
 * `\d` is any Unicode decimal digit, `\s` Python's whitespace, `\b` a
 * boundary between Python word characters.
 *
 * One difference: the script reads the JD from stdin when `--jd` is missing
 * or empty; a tool has no stdin, so that case is an empty JD — an empty line.
 */

import { parseArgv } from "../../db/argv.ts";
import { PY_SPACE_CLASS, pyInt } from "../../db/py-format.ts";
import type { ScriptResult } from "../../db/tools.ts";
import type { ToolHandler } from "../../tools/registry.ts";
import { argvTool } from "./argv-tool.ts";

/** `MONTHS`, in the dict's order: the alternation tries them in this order, as Python's does. */
const MONTHS: ReadonlyArray<[string, number]> = [
  ["jan", 1], ["january", 1], ["gen", 1], ["gennaio", 1],
  ["feb", 2], ["february", 2], ["febbraio", 2],
  ["mar", 3], ["march", 3], ["marzo", 3],
  ["apr", 4], ["april", 4], ["aprile", 4],
  ["may", 5], ["maggio", 5], ["mag", 5],
  ["jun", 6], ["june", 6], ["giu", 6], ["giugno", 6],
  ["jul", 7], ["july", 7], ["lug", 7], ["luglio", 7],
  ["aug", 8], ["august", 8], ["ago", 8], ["agosto", 8],
  ["sep", 9], ["sept", 9], ["september", 9], ["set", 9], ["settembre", 9],
  ["oct", 10], ["october", 10], ["ott", 10], ["ottobre", 10],
  ["nov", 11], ["november", 11], ["novembre", 11],
  ["dec", 12], ["december", 12], ["dic", 12], ["dicembre", 12],
];
const MONTH = new Map(MONTHS);

/** Python's `\w` in a `str` pattern: letters, numbers, underscore. */
const W = String.raw`[\p{L}\p{N}_]`;
/** Python's `\b`. */
const B = `(?:(?<=${W})(?!${W})|(?<!${W})(?=${W}))`;
const D = String.raw`\p{Nd}`;
const S = `[${PY_SPACE_CLASS}]`;
const ALT = MONTHS.map(([m]) => m).join("|");

const ISO = new RegExp(`${B}(${D}{4})-(${D}{2})-(${D}{2})${B}`, "u");
const NUMERIC = new RegExp(`${B}(${D}{1,2})[/\\-](${D}{1,2})[/\\-](${D}{2,4})${B}`, "u");
const MONTH_DAY = new RegExp(`${B}(${ALT})${S}+(${D}{1,2})(?:[,${PY_SPACE_CLASS}]+(${D}{4}))?${B}`, "u");
const DAY_MONTH = new RegExp(`${B}(${D}{1,2})${S}+(${ALT})(?:${S}+(${D}{4}))?${B}`, "u");
const IN_DAYS = new RegExp(`${B}(?:expires?|closes?|deadline|scade|chiude)${S}+(?:in${S}+|fra${S}+|tra${S}+|entro${S}+)?(${D}+)${S}+(?:days?|giorni)${B}`, "u");

/** A calendar day, compared and printed as `datetime.date` does. */
interface Day {
  y: number;
  m: number;
  d: number;
}

const key = (x: Day) => x.y * 10_000 + x.m * 100 + x.d;
const isoOf = (x: Day) => `${String(x.y).padStart(4, "0")}-${String(x.m).padStart(2, "0")}-${String(x.d).padStart(2, "0")}`;

/** `_iso`: a real date between years 1 and 9999, or null (`ValueError`). */
function day(y: number, m: number, d: number): Day | null {
  if (y < 1 || y > 9999 || m < 1 || m > 12 || d < 1) return null;
  const last = new Date(Date.UTC(2000, m, 0)).getUTCDate();
  const leap = y % 4 === 0 && (y % 100 !== 0 || y % 400 === 0);
  if (d > (m === 2 ? (leap ? 29 : 28) : last)) return null;
  return { y, m, d };
}

/** `today + timedelta(days=n)`, and Python's OverflowError past year 9999. */
function addDays(today: Day, n: number): Day {
  const t = Date.UTC(today.y, today.m - 1, today.d) + n * 86_400_000;
  const date = new Date(t);
  if (!Number.isFinite(t) || Number.isNaN(date.getTime()) || date.getUTCFullYear() > 9999) {
    throw new Error("date value out of range");
  }
  return { y: date.getUTCFullYear(), m: date.getUTCMonth() + 1, d: date.getUTCDate() };
}

/** `parse_deadline`, with today given (the script's is the local date). */
export function parseDeadline(text: string, today: Day): string | null {
  const t = text
    .replace(new RegExp(`${S}+`, "gu"), " ")
    .replace(new RegExp(`^${S}+|${S}+$`, "gu"), "")
    .toLowerCase();
  if (!t) return null;
  const int = (s: string) => pyInt(s)!;

  let m = ISO.exec(t);
  if (m) {
    const x = day(int(m[1]!), int(m[2]!), int(m[3]!));
    if (x && key(x) >= key(today)) return isoOf(x);
  }
  m = NUMERIC.exec(t);
  if (m) {
    let y = int(m[3]!);
    if (y < 100) y += 2000;
    const x = day(y, int(m[2]!), int(m[1]!));
    if (x && key(x) >= key(today)) return isoOf(x);
  }
  for (const [re, dayGroup, monthGroup] of [[MONTH_DAY, 2, 1], [DAY_MONTH, 1, 2]] as const) {
    m = re.exec(t);
    if (!m) continue;
    const mo = MONTH.get(m[monthGroup]!)!;
    const d = int(m[dayGroup]!);
    const y = m[3] ? int(m[3]) : today.y;
    let x = day(y, mo, d);
    if (x) {
      // No year and already past this year: next year's.
      if (!m[3] && key(x) < key(today)) x = day(y + 1, mo, d);
      if (x && key(x) >= key(today)) return isoOf(x);
    }
  }
  m = IN_DAYS.exec(t);
  if (m) {
    const digits = m[1]!;
    // More days than any date holds: the Python's timedelta overflows the same way.
    if (Array.from(digits).length > 9) throw new Error("date value out of range");
    return isoOf(addDays(today, int(digits)));
  }
  return null;
}

/** The local calendar day, as `datetime.now().date()`. */
function localToday(now: Date): Day {
  return { y: now.getFullYear(), m: now.getMonth() + 1, d: now.getDate() };
}

export function deadlineExtract(argv: string[], now: Date = new Date()): ScriptResult {
  const a = parseArgv({ prog: "deadline_extract.py", mainProg: "deadline_extract.py", options: [{ flag: "--jd" }] }, argv);
  const jd = typeof a["jd"] === "string" ? a["jd"] : "";
  return { stdout: `${parseDeadline(jd, localToday(now)) ?? ""}\n`, exitCode: 0 };
}

export function createDeadlineExtractTool(options: { now?: () => Date } = {}): ToolHandler {
  return argvTool({
    name: "deadline_extract",
    script: "deadline_extract.py",
    description: "Extract an application deadline from a JD's text (--jd \"...\"): an ISO date, or an empty line when none is certain.",
    run: (args) => deadlineExtract(args, options.now?.() ?? new Date()),
  });
}
